examples/systemcalls/systemcalls.c

```diff
@@ -6,94 +6,163 @@
+ *   The first is always the full path to the command to execute with execv()
+ *   The remaining arguments are a list of arguments to pass to the command in execv()
+ * @return true if the command @param ... with arguments @param arguments were executed successfully
+ *   using the execv() call, false if an error occurred, either in invocation of the
+ *   fork, waitpid, or execv() command, or if a non-zero return value was returned
+ *   by the command issued in @param arguments with the specified arguments.
+ */
 bool do_exec(int count, ...)
 {
     va_list args;
     va_start(args, count);
-    char * command[count+1];
+    char *command[count + 1];
     int i;
-    for(i=0; i<count; i++)
+    for (i = 0; i < count; i++)
     {
         command[i] = va_arg(args, char *);
     }
     command[count] = NULL;
-    // this line is to avoid a compile warning before your implementation is complete
-    // and may be removed
-    command[count] = command[count];
-
-/*
- * TODO:
- *   Execute a system command by calling fork, execv(),
- *   and wait instead of system (see LSP page 161).
- *   Use the command[0] as the full path to the command to execute
- *   (first argument to execv), and use the remaining arguments
- *   as second argument to the execv() command.
- *
-*/
 
     va_end(args);
 
-    return true;
+    /*
+     * TODO:
+     *   Execute a system command by calling fork, execv(),
+     *   and wait instead of system (see LSP page 161).
+     *   Use the command[0] as the full path to the command to execute
+     *   (first argument to execv), and use the remaining arguments
+     *   as second argument to the execv() command.
+     *
+     */
+    if (command[0][0] != '/') // Command must be an absolute path to binary
+    {
+        return false;
+    }
+
+    pid_t pid = fork();
+    if (pid == -1) // Failure
+    {
+        perror("fork failed!");
+        return false;
+    }
+
+    if (pid == 0) // Child process
+    {
+        execv(command[0], command);
+    }
+    // Else, parent process
+    /*
+     * waitpid() to communicate information about process' termination.
+     * This information is stored in status.
+     */
+    int status;
+    if (waitpid(pid, &status, 0) == -1)
+    {
+        perror("waitpid failed!");
+        return false;
+    }
+
+    // Checking if the child process exited normally AND if the exit code is 0 (success)
+    if (WIFEXITED(status) && WEXITSTATUS(status) == 0)
+    {
+        return true;
+    }
+    return false;
 }
 
 /**
-* @param outputfile - The full path to the file to write with command output.
-*   This file will be closed at completion of the function call.
-* All other parameters, see do_exec above
-*/
+ * @param outputfile - The full path to the file to write with command output.
+ *   This file will be closed at completion of the function call.
+ * All other parameters, see do_exec above
+ */
 bool do_exec_redirect(const char *outputfile, int count, ...)
```

Comment:
@Sriram-Y see discussion in class about how you could avoid duplicating logic here with the function above.